### DISCUSSION TITLE
API token with revocation

### DISCUSSION THREAD

**Discussion by leviia** [pocketbase/pocketbase#2900] [Show and tell]
Hello everyone 

It would be very useful to have a way to generate api token in poketbase to allow our customer to connect they tools to our backend. (e.g. PocketBase)

For this we need a way to generate token with custom duration and also a way for our user to revocate the token if needed.
We might be wrong but here is the function that generate tokens: https://github.com/pocketbase/pocketbase/blob/c3844250e807f7a8292f00f328e405c45fea6725/tokens/record.go#L13

```golang
func NewRecordAuthToken(app core.App, record *models.Record) (string, error) {
	if !record.Collection().IsAuth() {
		return "", errors.New("The record is not from an auth collection.")
	}

	return security.NewToken(
		jwt.MapClaims{
			"id":           record.Id,
			"type":         TypeAuthRecord,
			"collectionId": record.Collection().Id,
		},
		(record.TokenKey() + app.Settings().RecordAuthToken.Secret),
		app.Settings().RecordAuthToken.Duration,
	)
}
```

It might be interesting to have another function dedicated to api Tokens: 

```golang
func NewRecordAPIAuthToken(app core.App, record *models.Record, CustomDuration) (string, error) {
	if !record.Collection().IsAuth() {
		return "", errors.New("The record is not from an auth collection.")
	}

	token := security.NewToken(
		jwt.MapClaims{
			"id":           record.Id,
			"type":         TypeAPIAuthRecord,  // <-- switched from TypeAuthRecord to TypeAPIAuthRecord
			"collectionId": record.Collection().Id,
		},
		(record.TokenKey() + app.Settings().RecordAuthToken.Secret),
		CustomDuration,  // <-- Added custom duration variable
	)
	SaveTokenInDatabaseAPIToken(token) // <-- save before return so it becomes possible to check later
	return token
}
```
As the token is stored in a database it becomes possible to check this token at each connection via a middleware (don't know where yet). 

Something like this:

 ```
// At each connection
If token is  TypeAPIAuthRecord
    If token is in DatabaseAPIToken
      return ok 
    return Nok
Pass
```

Then we need some function like 

`GetAPIAuthToken(user)`
`RemoveAPIAuthToken(token)`

So the end user is able to generate api token with custom duration, is able to see them in the frontend and is also able to revocate them if needed.

What do you think of such a functionality? We can implement it if you agree and then make a push request.

All the best

Arnaud

**Comment by ganigeorgiev** [comment]
PocketBase API is fully stateless and there are no plans for now to store the tokens in the database (_keep in mind that storing them in the database will require some cleanup function to prevent the tokens table to grow indefinitely_).

Users are free to implement this behavior on their own if they really need it, but I don't think the use case is general enough to be part of core package.

Fixed Admin tokens or API keys may get added in the future (see https://github.com/pocketbase/pocketbase/discussions/342) but for now remains a low priority.

**Comment by leviia** [comment]
Ok thanks 

We will implement it on our own and if this is needed let us know we can make a push request

All the best 

Arnaud
